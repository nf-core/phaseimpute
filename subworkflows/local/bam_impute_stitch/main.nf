include { STITCH               } from '../../../modules/nf-core/stitch'
include { BCFTOOLS_INDEX       } from '../../../modules/nf-core/bcftools/index'

workflow BAM_IMPUTE_STITCH {

    take:
    ch_bam         // channel:   [ [id], bam, bai, bamlist ]
    ch_parameters  // channel:   [ [chr], posfile, input, rdata, chr, k_val, ngen]
    ch_fasta       // channel:   [ [genome], fa, fai ]

    main:

    ch_versions      = Channel.empty()
    // Run STITCH
    seed = params.seed

    ch_bam_params = ch_bam // Add chr to meta map
        .combine(ch_parameters)
        .map{
            metaI, bam, bai, bamlist, metaPC, posfile, input, rdata, chr, k_val, ngen ->
            [metaI + [chr: metaPC.chr, panel:metaPC.id], bam, bai, bamlist, posfile, input, rdata, chr, k_val, ngen]
        }

    STITCH( ch_bam_params, ch_fasta, seed )
    ch_versions = ch_versions.mix(STITCH.out.versions)

    // Index imputed annotated VCF
    BCFTOOLS_INDEX(STITCH.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions)

    // Join VCFs and TBIs
    ch_vcf_tbi = STITCH.out.vcf
        .join(BCFTOOLS_INDEX.out.tbi)
        .map { metaI, vcf, tbi -> [ metaI + [tools: "stitch"], vcf, tbi ] }

    emit:
    vcf_tbi  = ch_vcf_tbi                        // channel:   [ [id, chr], vcf, tbi ]
    versions = ch_versions                       // channel:   [ versions.yml ]

}
