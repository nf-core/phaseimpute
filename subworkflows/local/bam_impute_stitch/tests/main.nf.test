nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_STITCH"
    script "../main.nf"
    config "./nextflow.config"

    workflow "BAM_IMPUTE_STITCH"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_impute_stitch"
    tag "bam_impute_stitch"

    tag "stitch"
    tag "bcftools"
    tag "bcftools_index"

    test("Prepare input stitch") {
        setup {
            run("PREPARE_INPUT_STITCH") {
                script "../../prepare_input_stitch/main.nf"
                process {
                    """
                    input[0] = Channel.fromList([
                        [
                            [id: "NA12878"],
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                        ],
                        [
                            [id: "NA19401"],
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true),
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true),
                        ],
                    ])
                    input[1] = Channel.of([
                            [id: "1000GP", chr: "chr22"],
                            file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                        ]).collect()
                    input[2] = Channel.of([
                        [chr: "chr22", region: "chr22:16570000-16610000"], "chr22:16570000-16610000"
                    ]).collect()
                    """
                }
            }
        }
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = PREPARE_INPUT_STITCH.out.stitch_samples
                input[1] = PREPARE_INPUT_STITCH.out.stitch_parameters
                input[2] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary }).match("SummaryMixed") }
            )
        }
    }
}
