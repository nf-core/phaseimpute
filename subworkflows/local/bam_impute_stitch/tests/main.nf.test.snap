{
    "Prepare input stitch": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "all_samples",
                            "chr": "chr22",
                            "panel": "1000GP",
                            "tools": "stitch"
                        },
                        "stitch.chr22.vcf.gz:md5,9d3c27d9f0325cc92c22234665f5bb12",
                        "stitch.chr22.vcf.gz.tbi:md5,7d8d37b65e631c2b40c54e4a175de86c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,4330856a215c4986b0591d550a4d22bf",
                    "versions.yml:md5,8561197b98b9a1e1f8f1b01c1369da35"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "all_samples",
                            "chr": "chr22",
                            "panel": "1000GP",
                            "tools": "stitch"
                        },
                        "stitch.chr22.vcf.gz:md5,9d3c27d9f0325cc92c22234665f5bb12",
                        "stitch.chr22.vcf.gz.tbi:md5,7d8d37b65e631c2b40c54e4a175de86c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4330856a215c4986b0591d550a4d22bf",
                    "versions.yml:md5,8561197b98b9a1e1f8f1b01c1369da35"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:32:53.365697746"
    },
    "SummaryMixed": {
        "content": [
            [
                "VcfFile [chromosomes=[22], sampleCount=2, variantCount=903, phased=false, phasedAutodetect=false]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:32:53.598743638"
    }
}